Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -139,89 +140,70 @@ namespace ROS2
-                "y",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32,
-                "z",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32,
-                "intensity",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32);
+            builder.AddField("intensity", sensor_msgs::msg::PointField::FLOAT32);
         }
-        else
-        {
-            modifier.setPointCloud2Fields(
-                3,
-                "x",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32,
-                "y",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32,
-                "z",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32);
-        }
-        modifier.resize(results.m_points.size());
+        sensor_msgs::msg::PointCloud2 message = builder.Build();
+
+
+        sensor_msgs::PointCloud2Iterator<float> messageXIt(message, "x");
+        sensor_msgs::PointCloud2Iterator<float> messageYIt(message, "y");
+        sensor_msgs::PointCloud2Iterator<float> messageZIt(message, "z");
 
-        sensor_msgs::PointCloud2Iterator<float> xIt(message, "x");
-        sensor_msgs::PointCloud2Iterator<float> yIt(message, "y");
-        sensor_msgs::PointCloud2Iterator<float> zIt(message, "z");
+        const auto positionField = results.GetConstFieldSpan<RaycastResultFlags::Point>().value();
+        auto positionIt = positionField.begin();
 
-        AZStd::optional<sensor_msgs::PointCloud2Iterator<float>> intensityIt;
+        AZStd::optional<sensor_msgs::PointCloud2Iterator<float>> messageIntensityIt;
+        AZStd::optional<RaycastResults::FieldSpan<RaycastResultFlags::Intensity>::const_iterator> intensityIt;
         if (isIntensityEnabled)
         {
-            intensityIt = sensor_msgs::PointCloud2Iterator<float>(message, "intensity");
+            messageIntensityIt = sensor_msgs::PointCloud2Iterator<float>(message, "intensity");
+            intensityIt = results.GetConstFieldSpan<RaycastResultFlags::Intensity>().value().begin();
         }
 
         const auto entityTransform = GetEntity()->FindComponent<AzFramework::TransformComponent>();
         const auto inverseLidarTM = entityTransform->GetWorldTM().GetInverse();
-        for (size_t i = 0; i < results.m_points.size(); ++i)
+        for (size_t i = 0; i < results.GetCount(); ++i)
         {
-            const AZ::Vector3 globalPoint = inverseLidarTM.TransformPoint(results.m_points[i]);
-            *xIt = globalPoint.GetX();
-            *yIt = globalPoint.GetY();
-            *zIt = globalPoint.GetZ();
+            AZ::Vector3 point = *positionIt;
+            const AZ::Vector3 globalPoint = inverseLidarTM.TransformPoint(point);
+            *messageXIt = globalPoint.GetX();
+            *messageYIt = globalPoint.GetY();
+            *messageZIt = globalPoint.GetZ();
 
             if (isIntensityEnabled)
```

Comment:
I wonder why 'isIntensityEnabled' is used instead of messageIntensityIt that is optional?
Maybe no need wrap iterator in optional type?

Reply:
The iterator has to be wrapped in optional since to create it, intensity field has to be present in both the message and the raycast results. Additionally the presence of both iterators is synonymous with `isIntensityEnabled`. Since however someone else has had a similar insight before, I will change this to use `if (messageIntensityIt.has_value() && intensityIt.has_value())` instead.